### PR TITLE
Serialization/Deserialization

### LINKED ISSUE #2: Provide serialization/deserialization mechanisms

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Resolves #2 

@iravid Still a WIP but just as a sanity check if you want to take a look. This is my last week at the G so I'll be able to devote more time going forward.

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
I always forget to hit the "Draft PR"  button, GH should really unhide it.

File: src/main/scala/zio/rocksdb/Serializer.scala [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+private[rocksdb] trait SerializerUtilityFunctions {
+  private[rocksdb] def fromByteBuffer[A](n0: Int, mutate: (ByteBuffer, A) => Any): Serializer[Any, A] =
+    new Serializer[Any, A] {
+      var buf: ByteBuffer = null
+      def apply(a: A): UIO[Bytes] = {
+        val n = n0 max 1
+        for {
+          _ <- URIO.effectTotal { buf = ByteBuffer.allocate(n) }.when(buf eq null)
```
Comment (iravid) [inline src/main/scala/zio/rocksdb/Serializer.scala]: Maybe simplify to this?
```scala
UIO {
  if (buf eq null)
    buf = ByteBuffer.allocate(n)
  buf.position(0)
  mutate(buf, a)
  Chunk.fromArray(buf.array)
}
```

File: src/main/scala/zio/rocksdb/Serializer.scala [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+trait Serializer[-R, -A] { self =>
+  def apply(a: A): URIO[R, Bytes]
+
+  final def contramap[B](f: B => A): Serializer[R, B] =
+    new Serializer[R, B] {
+      def apply(b: B): URIO[R, Bytes] = self(f(b))
+    }
+
+  final def zipWith[B, C](f: B => (A, C))(that: Serializer[R, C]): Serializer[R, B] =
```
Comment (iravid) [inline src/main/scala/zio/rocksdb/Serializer.scala]:
```suggestion
  final def zipWith[B, C](that: Serializer[R, C])(f: B => (A, C)): Serializer[R, B] =
```

File: src/main/scala/zio/rocksdb/Serializer.scala [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    new Serializer[R, B] {
+      def apply(b: B): URIO[R, Bytes] = {
+        val (a, c) = f(b)
+        self(a).zipWith(that(c))(_ ++ _)
+      }
+    }
+
+  final def zip[B](that: Serializer[R, B]): Serializer[R, (A, B)] = zipWith(identity[(A, B)])(that)
+
+  final def chooseWith[B, C](f: B => Either[A, C])(that: Serializer[R, C]): Serializer[R, B] =
```
Comment (iravid) [inline src/main/scala/zio/rocksdb/Serializer.scala]:
```suggestion
  final def chooseWith[B, C](that: Serializer[R, C])(f: B => Either[A, C]): Serializer[R, B] =
```

File: src/main/scala/zio/rocksdb/Serializer.scala (hunk 2 above)
Comment (iravid) [reply src/main/scala/zio/rocksdb/Serializer.scala]: Also, should we call it `divideWith`?

**Review by iravid — COMMENTED** [review]
Looks great so far @regiskuckaertz!

File: src/main/scala/zio/rocksdb/Serializer.scala (hunk 2 above)
Comment (regiskuckaertz) [reply src/main/scala/zio/rocksdb/Serializer.scala]: Yep, that was the original name 😅

File: src/main/scala/zio/rocksdb/Deserializer.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package zio.rocksdb
+
+import zio.ZIO
+
+trait Deserializer[-R, +A] { self =>
+  def decode(bytes: Bytes): ZIO[R, DeserializeError, Result[A]]
+
+  final def map[B, A1 >: A](f: A1 => B): Deserializer[R, B] =
```
Comment (iravid) [inline src/main/scala/zio/rocksdb/Deserializer.scala]: Is `A1` needed here? It's usually not a problematic position for covariant type variables

Comment (regiskuckaertz) [reply src/main/scala/zio/rocksdb/Deserializer.scala]: Yup, I thought so but you are right.

**Comment by iravid** [comment]
@regiskuckaertz Let me know when this is ready for review!

**Comment by regiskuckaertz** [comment]
@iravid yeah I think this is ready. There are still many tests I'll add but they won't diverge from what is already there.

**Review by iravid — APPROVED** [review]
@regiskuckaertz Great work and a massive undertaking! Looks excellent. As a follow-up, we may want to consider parameterizing `RocksDB` by `K` and `V` and baking the `Serde` into that class.

**Comment by iravid** [comment]
@regiskuckaertz Thanks for picking this up!

**Pull request merged**